redfish-core/lib/account_service.hpp

```diff
@@ -1997,23 +2130,49 @@ inline void requestAccountServiceRoutes(App& app)
             std::optional<bool> locked;
             std::optional<nlohmann::json> oem;
 
+            if (!json_util::readJson(req, asyncResp->res, "UserName",
+                                     newUserName, "Password", password,
+                                     "RoleId", roleId, "Enabled", enabled,
+                                     "Locked", locked, "Oem", oem))
+            {
+                return;
+            }
+
+            // Unauthenticated user
+            if (req.session == nullptr)
+            {
+                // If user is service
+                if (username == "service")
+                {
+                    if (oem)
+                    {
+                        // allow unauthenticated ACF upload based on panel
+                        // function 74 state.
+                        triggerUnauthenticatedACFUpload(asyncResp, oem);
+                        return;
+                    }
+                }
+                messages::insufficientPrivilege(asyncResp->res);
+                return;
+            }
+
             Privileges effectiveUserPrivileges =
                 redfish::getUserPrivileges(req.userRole);
             Privileges configureUsers = {"ConfigureUsers"};
             bool userHasConfigureUsers =
                 effectiveUserPrivileges.isSupersetOf(configureUsers);
-            if (userHasConfigureUsers)
+            if (!userHasConfigureUsers)
             {
-                if (!json_util::readJson(req, asyncResp->res, "UserName",
-                                         newUserName, "Password", password,
-                                         "RoleId", roleId, "Enabled", enabled,
-                                         "Locked", locked, "Oem", oem))
+                // Irrespective of role can patch ACF if function
+                // 74 is active from panel.
+                if (oem && (username == "service"))
```

Comment:
The only one without a ConfigureUsers would be an operator  ?
Why do we need this? Why can't we just handle this normally if the user has authenticated ?

Reply:
I am not sure if username other than service should be allowed to update acf certificates. As we see in the existing code (line 2073) even an authenticated user is allowed to upload acf certificate only if username is "Service".

https://github.com/ibm-openbmc/bmcweb/pull/243/files#diff-a6441dc86bbef5cbe2192c66717e8b938f9ee3a02670ff54c702338095a0d660L2073

Reply:
Correction, this would be operatior, readonly, and noaccess user. I am still not sure why we need this but I'll default to joseph here

Reply:
I met with Sunny to discuss this.  We followed the code flows for the 3 conditions:
> (((user is not authenticated) && (isACFWindowActive==true)) ||
> ((!userHasConfigureUsers) && (isACFWindowActive==true)) ||
> userHasConfigureUsers)

The first two conditions were easy to validate.  For the third condition, we walked through the logic in 2164-2188 and through several PATCH requests:
- Line 2168: PATCHing {"Oem":{"IBM":{"ACF": ... }}}.
These two clauses are unchanged:
- Line 2177: PATCHing someone else's account (username != req.session->username).
- Line 2185: PATCHing just the RoleId, or PATCHing both "Password" and "RoleId" in the same request.

Gunnar, to address your question directly: I'm not sure what you mean by "normally".  The new code follows the logic from old line 2005 else clause (lines 2015-2029).  The old code allowed a user who does not have ConfigureUsers to change their own account password.  The new code allows that, plus it allows uploading an ACF when isACFWindowActive==true.  That looks right to me.